### ISSUE TITLE
Create a Metrics Manager

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#1340]
**Is your feature request related to a problem? Please describe.**
In many separate places metric over time are captured using a variety of classes. The individual implementations have different retention periods and data storage.
Viewing the metrics is custom to each one, needing bespoke code in the UI.
Furthermore there is no ability to retain the data for long periods. 
Relates to #973

**Describe the rationale**
Metrics is key to understand how the Simulation and Settlement are performing.

**Describe the goal(s) you have in mind**
Consolidate the various metrics into a single solution that provides a consist and universal way to store any numerical metric. Having a single solution will lead to a sharable means to render the metrics using tabular or chart based presentations. These could show changes over time.
A central solution will also allow better management of the storage of these metrics which may become large

**Describe the solution you desire to see**
Create a new ```MetricManager``` class that is the central point for holding metrics. An instance is created per Settlement which removes the problem of multiple updaters. This will replace all '''SolLogger''' classes.
The Metric manager will capture absolute value of a named metric at a point in time. It will be able to retrieve the sequence of metrics over time. Metrics will be identified by 2 attributes:

1.  Type - This could be Resources Consumed, Active Missions, EVA hours
2.  Instance - This will be a sub-type and specific to the Type. For example, name of resource for Consumed, Type of Mission or Person name

The attributes will be String based to allow for expansion.
The Metric Manager will also handle storage by writing old sols to file as CSV ? to control the amount of memory consumed.
Older data will automatically be loaded from file store if it is required; this will use the Future concept.

To accompany this, a central UI will be provided that can chart any selected metric over time allowing metrics from different type to be rendered side by side.

**Comment by mokun** [comment]
Very good idea to have a unified Metric Manager that can output to CSV at player's demand by pressing a button on a UI or simply on console. 

btw, it's actually good for us developers to be able to do a data dump on, say, the production or consumption of a methane or any resources; or the power/heat consumed or generated at a power/heat source of building, etc.

**Comment by bevans2000** [comment]
It wasn't in my original idea to have an export to CSV but we can add it.
The idea was to use CSVs to hold old data so less is kept in memory.

**Comment by mokun** [comment]
How do you like to handle saving and loading from external files ?

Creating a default.csv or default.json for the data ?

**Comment by bevans2000** [comment]
For this task I was thinking plain old CSV because it will be aggregated counts over different metrics. As the data is homogeneous in structure, CSV makes sense. If there was a lot of variance then JSON would be the choice but I see this as very regular.
Do you think we should do it for 3.9?
We could create the manager and then have a set of follow up activities to use it in various places. Probably do it as Project so we can track the various parts

**Comment by bevans2000** [comment]
We don't need to serialise the historical data. We can save it to file as CSV. There is no need to use JSON.

**Comment by bevans2000** [comment]
I see this being the properties of any measurement
- time -when the measurement occurs
- metric - what is being measured, example, resource consumed, number of active missions, time spent on Eva
- category - metrics have sub category if needed. Would have a default value of 'all' but  example would be the type of resource consumed
- entity - Entity where the measurement took place

Measurement of the current sil are held in memory. Older measurements are flushed to the file system to reduce memory consumption and save simulation size.
The Metro manager will automatically load and cache old metric loaded from file.

**Comment by mokun** [comment]
In some cases, all I need is just one daily average value.

For instance, I want to know the average power usage of a vehicle whenever it's on the road.

Say the average value is around 35 kW on sol 2. 30 kW on sol 3. 40 kW on sol 4. The final average out of these 3 sols is 30 kW.

Will this Metrics Manager be enable to plot a graph for me. In this case, x is the sol, y is the average power usage ?

**Comment by bevans2000** [comment]
Yes we can work it to do that. Probably record one value per sol for long term storage

**Comment by mokun** [comment]
> Probably record one value per sol for long term storage

In terms of that example about vehicle power usage, it does capture, say, 200 power usage values (10 kW, 15 kW, 45 kW, 30 kW, 35 kW, 20 kW, 22 kW, 30 kW....) 

But at the end of that sol, all I'm interested in is ONE SINGLE average value computed from those 200 values.

So once the daily average is being computed and is saved, we can safely delete those 200 values captured on the previous sol.

This is the convoluted part.

**Comment by bevans2000** [comment]
Yes daily average may need an intermediate in memory structure, probably what we have now.
But the purpose behind this proposal was the ask to track the consumption and creation of resources over time so we can see any trends. This cannot be held in memory all time because it will occupy too much.

**Comment by bevans2000** [comment]
Another way to look at this is we record the individual energy consumption by time and then once a sol calculate the average. How is the daily average used?

**Comment by mokun** [comment]
I just came across the idea of MELT (metrics, events, logs, traces).

It's been the default strategy for understanding network health and security.

It promises visibility into records of execution, like a digital diary, to empower tools with a historical view of what actions occurred. 

We could empower all 4 aspects in mars-sim.

**Comment by bevans2000** [comment]
Yes I know of these principles. I have some strong thoughts on this now and a rough design. I will try and write them up before igoon holiday

**Comment by bevans2000** [comment]
Exactly. I already have thought it out. The data saved to files should be selectively reloaded to populate any charts as well. This would be automatic

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

So for this metric manager, I'd like to see if we can incorporate more of their methods from 3 of those [apache statistics libraries](https://commons.apache.org/proper/commons-statistics/) (which I've just added them recently and they've got new fixes and v1.2 for us to [pull](https://github.com/mars-sim/mars-sim/pull/1752)) as we don't want to waste time reinventing the wheel if most methods are already out there.

No doubt we do need to customize them for our use cases.

**Comment by bevans2000** [comment]
Ok, sounds good. I've got the design sketched out but I don't want to start this yet. We need to finish what we have on the go in 3.10

**Comment by mokun** [comment]
> do we also track any kind of emissions of our settlements or rovers as statistic? (or better, as accidental terraforming)

Ha ! Robert Zubrin wants to start a greenhouse effect on Mars to warm up Mars. Since Mars atmosphere has already been filled with CO2, I suppose he doesn't care about methane emissions :D and praise Musk's decision to develop methane powered engines...

**Comment by mokun** [comment]
@Urwumpe 

Back to the question, currently, we did track how much O2 and CO2 being released/consumed by food crops in every greenhouse. We could get a total from a settlement. 

It will be nice to develop some UI fro this Metric manager (I imagine the Simcity and Cities II: Skylines also have graph manager) that player can have access for viewing these statistics.

**Comment by Urwumpe** [comment]
I think the O2 and CO2 of greenhouses is still kept inside the settlement, but not released to contaminate Mars. Yes, its a terraforming question, but also a process to quickly reduce inventory sizes.

**Comment by bevans2000** [comment]
For this issue I just want to improve how we capture various metrics in a generic way. At the moment we use bespoke specific approaches which makes it difficult to use them. In theory, if we have a generic way it could be used for the Simulation goals. It would be possible to specify a goal based on a metric.

**Comment by bevans2000** [comment]
Here are some thoughts on how this can be implemented.
Metrics are a sequence of time based measurements or data points. They are unique under a number of attributes:
- Unit Measured - ```Entity```, e.g. could be a ```Mission```, ```Settlement``` etc etc
- Category- The classification of the the thing being measured; String based, e.g. population, production, resource level
- Measure - The actual property measured e.g, varies but is String based, Oxygen, Process instance, etc

Controlled by a single manager but Metrics are grouped by Entity to avoid concurrency issues. The Metrics can be filtered/searched on Entity & Category & drill down into the Measure selected

Every data point is automatically timestamped. Any object can then record a value against a keyed Measure, e.g. number of Missions approved or Oxygen produced.
Store rolls over each Sol. Old sols are archived out to disk

Metrics support Visitor pattern which takes a ```Calculator``` interface instance. Can be applied to the all measures in a Metric or groups by Sol.

![image](https://www.plantuml.com/plantuml/png/fLNDRjim3BxhAOYSqit9xa4GT6jNeDWo34tHkshJZW15CYGwdR7jtKTQJcolQ0xRmQXAVlpuprn7qe6QWrMvrJ725YcO_0ksy5C1V7Pag8KLi0g91JlMkee5kIQiV6Z7iWFgs0HKlrLFjlDsXy5WxCZ0E69Kz7P0t6hItpqh4cAhGtmq1sJOonwTY0hVF5iH7BLjEcv6OoXrZd2ZRTvOJJvqvaTl2j1vZZND1-BExcfm-V68GLV85WDxXTIBvu8yGHQBNvkHarE9ZDlg7t-Piv1D9o86TMm9_ufCEKy85ai2NuuLjwOiCQ0ZC8VQue5Vcent4SWBBnHdGC3OM8fBXQu0uLEJDdOv3ipInxgsxNmaGYcFl2JuhQxNzuwoGRz95fzQRkNSE9BkFMXNOGPtZSibh5SMt1h-sqhgqnnJKAlLx48KCtL_IZg2XYSJZG1gJOJ1WUOynLfcYlQJ2I3FD1XQA3HfgANtGrak1V9FoVr_d6eOqIhufiO2dbjPcGpC4fTGUclzIsS-BD9ItQK080-e2SMtmvSnVyaVwdmlHb2Oc7lsXiLIFI0rmOb5pjJOvJSTlPY1TmWr1Z5UgaTslJQRNL-MzGUpOLj7mLlUJyC-k6BTM1ub_mmMH-pYjJeIV0EEv3rBDCGQSrEQN6ZV6jQjThh2jCbVJQHrkYiRMCdGtwHxOp3Eqr5A9-dg14x27jpz8z14SJvMWaxE4YwzUy8CUX_ysgPB9dXT54zoGTv6PMddnpT65kbyIRFqsd1c34YJ64orbr6SviuJUI7LsT11Qs50GHFJx8pgd0n7bRxizLBDfcvxc3WLrDHh7zFOxxItYyKxgOtYBxqxRGzYoMRu_JaB6JOxQDVEXYrGvzTOVTudTSt-cuFz0m00)

**Comment by mokun** [comment]
> Every data point is automatically timestamped. Any object can then record a value against a keyed Measure,

Yes. What we want is to be able to see these 2-D data points visually.

In terms of Simcity, players would like to see how population rises and falls over time.

**Comment by mokun** [comment]
In terms of storage, we may use [Java Parquet](https://www.baeldung.com/java-intro-to-apache-parquet). it's a columnar storage scheme and is perfect for fast reading the same kind of data while limiting the I/O.


My Brave AI said this : 
```
CSV remains a row-based storage format, meaning data is stored sequentially by rows, with each row representing a complete record.

This contrasts with Parquet, which is a columnar storage format designed to store data by columns rather than rows.
 In a row-based format like CSV, accessing a single column requires reading the entire row, which can be inefficient for large datasets.

Conversely, Parquet's columnar structure allows for independent access to each column, enabling more efficient data retrieval, especially when only specific columns are needed for analysis.

This fundamental difference in storage architecture leads to significant advantages for Parquet in terms of compression, query performance, and storage efficiency, particularly for large-scale data analytics.
```

**Comment by bevans2000** [comment]
Columnar storage is not needed for this repository. There is no need to return a subset of columns in this use case. A simple  RDMS will do it.

**Comment by Urwumpe** [comment]
In the worst case, we can use RDMS with a snowflake schema, to quickly work with multiple data sets. If the need arises. But for a Simcity like history of different plots, the plain RDMS is likely the easiest approach.

**Comment by Urwumpe** [comment]
In-Memory is OK for me, not sure if I even want to have a persistent detailled record. I would even like to have an option of how many sols or mars orbits / stellar years should be kept in DB and what can be truncated away. I don't see much sense in keeping fine-grained statistics for long periods of time. 

For long term-record keeping, I would rather like to see a summary report in HTML that is written on disc as well (for sharing). Maybe even starting with some highlights of what happened in each base during the past year. And some orbituaries for the deceased astronauts. Something that adds more value to the data and encourages sharing and engaging in a community of mars-sim players.

(Maybe just use XML or JSON for the data and add an external "prettifier" for making the data human readable.... and then create a web site on which people can share their last annual or 10 year summary report to the world.... And then paid subscription for a premium service ...)

**Comment by mokun** [comment]
Q: do we continue to support the graphing and charting capability in Monitor Tool ?

**Comment by mokun** [comment]
> add a generic UI panel (or pinch one or get Copilot to write me one)

Which library will you use ?

**Comment by bevans2000** [comment]
> Q: do we continue to support the graphing and charting capability in Monitor Tool ?
> 
> 
> 
Well it could go either way 

The charting in monitor can chart any attribute from the monitor tab, they do not have to be explicitly recorded. These are captured over time. Close the chart the data is lost.

The new charting tool will only show metrics that are explicitly captured. And will shown them over time.

However it seems wrong to have two charting so I'm in favour of dropping the Insitubl charting and only have a single approach. The most common numbers used from the monitor chart we can capture as explicitly Metrics.

But that's a way off yet

**Comment by mokun** [comment]
btw, since we are most likely deal with time series data, it's good to integrate the use of those 3 Apache Commons Statistics libraries into its methods.

```xml
			<groupId>org.apache.commons</groupId>
			<artifactId>commons-statistics-descriptive</artifactId>
			<version>1.2</version>
		</dependency>
		<dependency>
			<groupId>org.apache.commons</groupId>
			<artifactId>commons-statistics-distribution</artifactId>
			<version>1.2</version>
		</dependency>
		<dependency>
			<groupId>org.apache.commons</groupId>
			<artifactId>commons-statistics-inference</artifactId>
			<version>1.2</version>
		</dependency>	
```

When I ask Brave browser AI, it said the followings : 

- The Apache Commons Statistics library provides tools for statistical analysis, including functions relevant to time series data such as computing covariances and correlations between arrays or matrix columns.

- It includes packages for calculating Pearson's, Spearman's, and Kendall's correlation coefficients, which are commonly used in time series analysis to measure relationships between variables over time.

- While the library offers robust statistical functionality, it does not specifically focus on time series data management or analysis features like data ingestion, storage, or query optimization, which are more characteristic of dedicated time series databases.

- Therefore, while Apache Commons Statistics can support certain analytical tasks on time series data, it is not a dedicated time series database or data processing framework.

**Comment by mokun** [comment]
> Don't they just want to see how the values change over time

Yes i'd say seeing is believing. Players define want to see visually simcity-like graphs. It's the most basic thing.

Some of us want more beyond that.

Also, in order to rank settlements in terms of achievement, we'll need to compare an ensemble of numbers in a statistical manner. 

That's when we do need these statistical tools.

**Comment by bevans2000** [comment]
@mokun The way I want to prove this works is by showing metric for Missions. But the single `MissionManager` does not lead itself easily to this approach. So I'm thinking of creating a Settlement specific `MissionControl` that has most of the central manager but controls just Missions for a Settlement.
There will still be the need for an overarching manager but it will be a read only orchestrator class.
The will be a performance improvement as searches will already be scoped by a Settlement. Also it reduces a major contention point.
We could expand this to be one MissionControl per Authority in the future.

What do you think? The new MissionControl would be perfectly aligned to create metrics

**Comment by mokun** [comment]
> I'm thinking of creating a Settlement specific MissionControl that has most of the central manager but controls just Missions for a Settlement.

But what does the mission control have to do with the statistics of a settlement ?

**Comment by mokun** [comment]
For instance, I want to see how the amount of the oxygen rises and fall in a settlement.

I only have to select the settlement and the oxygen under resource.

I suppose now you're saying we could select not just a particular settlement but possibly a sponsor down the road, right ?

That would be fabulous !

I imagine when 3 nearby settlements join together governmentally, to, say, form a conglomerate, then it will already have the built-in flexibility available for players to choose that conglomerate and see how much combined resources they may have.

**Comment by bevans2000** [comment]
> > I'm thinking of creating a Settlement specific MissionControl that has most of the central manager but controls just Missions for a Settlement.
> 
> But what does the mission control have to do with the statistics of a settlement ? 

It doesn't actually. I was focusing on Missions as the example of how to use Metrics. But your comment on oxygen levels is a good point. 
I'll create an issue describing the Mission control idea but focus on capturing key resources in a Settlement. This will mean solving the memory consumption problem which I think can do.

**Comment by bevans2000** [comment]
Hmmmm, uncovered an interesting problem. I want to the large volume of Metric Data points in a dB so not to consume large qualities of memory, perfect answer. Can reuse loads of Spring data libraries.
The Metric is partially keyed on anything that implements an `Entity` so a reference to this has to be stored in the database table. But what is stored? The name well that's fragile and it does not connect back to the appropriate Entity object in memory. 
I've known this issue exists for a while because it is also a blocker for an API . We need to implement some form of unique identifier pattern for any Entity. This is basically an expansion of the Unit.identier but more general. This would then lead to an `EntityResolver` class. Until we can fix this identification problem everything will always have to sit in memory.
Doing your oxygen consumption tracking is easy now but it will blow away the memory eventually

**Comment by mokun** [comment]
So, is the alpha version of the metric manager with charting capability ready ? 

Can you commit that to Master branch so that I can see ?

I'm playing with Sector Base 1 with a population of 216 settlers. 

It will handy for me to see the trends in various life support resources. 

Currently, I'll have to take screen print of the Monitor Tool (or writing down in a piece of paper) for tracking resources.

**Comment by bevans2000** [comment]
Yes it's working but only monitoring a single metric. I've realised if we expanded it in the current approach it will blow the memory. 
I plan to store all the data points into a database to reduce memory footprint.
I needed the EntityResolver before doing it. Once the mission control is done, that will be the first test of it working at scale

**Comment by bevans2000** [comment]
Got a very basic POC working that stores Metrics in a H2 embedded database. Got CoPilot to do the coding.
Got to work how to backup and restore the database at start time to be synchronized with loading a saved database.

**Comment by mokun** [comment]
Can you commit this basic part ?

**Comment by mokun** [comment]
You use H2 database, right ? 

We used to load its maven artifacts and even try using it to save some data in Person.

But we didn't continue down that path to replace the standard Java serialization.

Now that you are using it in Metric manager.

Do you think there is a chance we expand it's use further ?

**Comment by bevans2000** [comment]
Indeed I think we could expand the use of a database further and stop serialization altogether.

I have just got the POC integrated into the new ```MissionControl``` class. To be honest that was the reason I did the mission control concept.
I will push something this weekend.

**Comment by mokun** [comment]
Hmm in what ways that metric manager is being used for mission statistics ?

**Comment by bevans2000** [comment]
Replacing and expanding the current SolHistory metrics. So now it will show the approved, denied, started and completed mission in one chart

**Comment by bevans2000** [comment]
An improvement to this is the```MemoryMetricsManager``` purges data points based on the consumer memory not the sol limit. This should be set as a percentage of the max heap so if a lot of memory is available, the metrics are allowed to grow more.
Also Metrics should start to replace the ```SolHistory``` instances because the former is more efficient.

**Comment by mokun** [comment]
Great job ! We now have a Metric Chart Viewer to visualize what's going on with missions. 

To me, it's essential for any mayors of a settlement and not just another toy. 

So I try that in my settlement Foundation below :  

<img width="1505" height="695" alt="Image" src="https://github.com/user-attachments/assets/7667a1b2-4f32-46bd-ad12-92cf79fe8acd" />

I'm very surprised to find out that on sol 8, it has attempted to kick start 340+ missions and close to 40 of them are "Successful". 

But according to the Monitor Tool's Mission (Note: it was mislabeled as "Buildings" at the title), 
19 out of 25 of them have 'truly' be completed.
 
<img width="1366" height="796" alt="Image" src="https://github.com/user-attachments/assets/d1208306-fefc-4965-b782-4db5e92b6f68" />

There are some Array OOB Exceptions when I click on the blue "Successful" curve.
```
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
```

I know we'll get rid of little bugs here and there one by one.

**Comment by bevans2000** [comment]
Great image of a lot of Missions in the chart. I feel it's about 80% there but yes expect some problem will need ironing out. But I wanted to give you the opportunity to try your Greenhouse idea

**Comment by mokun** [comment]
> the opportunity to try your Greenhouse idea

But I can't choose a building yet.

Also, for a single settlement template, I can't choose that settlement at all.

<img width="695" height="136" alt="Image" src="https://github.com/user-attachments/assets/b6856425-3f78-4333-8824-c762401bb5f2" />

The Chart Controls show "-- Select --" but with no settlement to choose from.

Can you take a look ?

**Comment by bevans2000** [comment]
It doesn't show until the Entity has a Metric. So you need to change the code to start logging a Metric for the Greenhouses.

**Comment by mokun** [comment]
I'm using a single settlement template.

Before I can see missions, I have to choose a settlement.

But I can't choose any settlement at all.

**Comment by bevans2000** [comment]
Not sure. It was working for me fine. The Settlement will only appear if the Metric has a value. The Metrics are not preregistered, they are created on the first data recorded.
But you had it working because you shared a screenshot

**Comment by mokun** [comment]
> But you had it working because you shared a screenshot

Earlier, it was working only in the default scenario when there are 18 settlements. I chose Foundation.

But I've quit that session and start a new one. This time I choose a template with just one settlement. 

Surprisingly, the metric viewer doesn't work.

**Comment by mokun** [comment]
> This time I choose a template with just one settlement.

Hmm now after I fast forward a few sol, missions are getting started on this small settlement.

Now I can select the settlement and see the mission stats.

**Comment by mokun** [comment]
I receive this below. It happens for the first time. So it's very rare.

No idea why having ClassCastException.

What output can it possibly trigger this class exception ?

```
01-Adir-04:423.484(Info) ReviewMissionPlan : [New Seoul] Do-yun Geo - Reviewing Exploration 3-2 - Rels: 4, Site: 84.4, Review: 2; Subtotal: 90.4
01-Adir-04:423.484(Info) ReviewMissionPlan : [New Seoul] Do-yun Geo - Done reviewing Eunu Lin's Exploration 3-2 mission plan.
01-Adir-04:424.252(Info) EVAOperation [x117] : [4.7962 S 0.0191 E - Gemini] Betty Ward - Transferring Specimen Box 777 from person back to rover.
01-Adir-04:424.252(Info) EVAOperation : [4.7962 S 0.0191 E - Gemini] Betty Ward - Done unloading all resources from person back to rover.
01-Adir-04:424.252(Info) ExploreSite [x59] : [4.7962 S 0.0191 E - Gemini] Betty Ward - Expected to collect 22.9 kg rocks to a specimen box.
01-Adir-04:424.252(Info) Exploration [x62] : [4.7962 S 0.0191 E - Gemini] Betty Ward - Ready to explore site and collect rocks.
01-Adir-04:425.795(Severe) TemporalThreadExecutor : Von-Braun: Problem with Pulse
java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1995)
	at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:2111)
	at java.base/java.util.HashMap.treeifyBin(HashMap.java:779)
	at java.base/java.util.HashMap.putVal(HashMap.java:651)
	at java.base/java.util.HashMap.put(HashMap.java:619)
	at com.mars_sim.core.metrics.memory.MemoryMetricManager.getMetric(MemoryMetricManager.java:170)
	at com.mars_sim.core.metrics.MetricManager.getMetric(MetricManager.java:70)
	at com.mars_sim.core.metrics.MetricGroup.lambda$0(MetricGroup.java:39)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1229)
	at com.mars_sim.core.metrics.MetricGroup.recordValue(MetricGroup.java:38)
	at com.mars_sim.core.structure.Settlement.addOutput(Settlement.java:2858)
	at com.mars_sim.core.process.ProcessItem.deposit(ProcessItem.java:147)
	at com.mars_sim.core.process.ProcessInfo.depositOutputs(ProcessInfo.java:213)
	at com.mars_sim.core.manufacture.WorkshopProcess.depositOutputs(WorkshopProcess.java:170)
	at com.mars_sim.core.manufacture.ManufactureProcess.stopProcess(ManufactureProcess.java:54)
	at com.mars_sim.core.manufacture.WorkshopProcess.checkStillActive(WorkshopProcess.java:127)
	at com.mars_sim.core.manufacture.WorkshopProcess.addWorkTime(WorkshopProcess.java:75)
	at com.mars_sim.core.manufacture.task.ManufactureWorkTask.workInWorkshop(ManufactureWorkTask.java:171)
	at com.mars_sim.core.manufacture.task.ManufactureWorkTask.manufacturePhase(ManufactureWorkTask.java:141)
	at com.mars_sim.core.manufacture.task.ManufactureWorkTask.performMappedPhase(ManufactureWorkTask.java:102)
	at com.mars_sim.core.person.ai.task.util.Task.executeMappedPhase(Task.java:653)
	at com.mars_sim.core.person.ai.task.util.Task.performTask(Task.java:622)
	at com.mars_sim.core.person.ai.task.util.TaskManager.executeTask(TaskManager.java:501)
	at com.mars_sim.core.robot.ai.BotMind.takeAction(BotMind.java:139)
	at com.mars_sim.core.robot.ai.BotMind.moderateTime(BotMind.java:108)
	at com.mars_sim.core.robot.ai.BotMind.timePassing(BotMind.java:87)
	at com.mars_sim.core.robot.Robot.timePassing(Robot.java:429)
	at com.mars_sim.core.structure.Settlement.timePassing(Settlement.java:1087)
	at com.mars_sim.core.structure.Settlement.timePassing(Settlement.java:991)
	at com.mars_sim.core.unit.TemporalThreadExecutor$TemporalRunnable.run(TemporalThreadExecutor.java:81)
	at java.base/java.lang.Thread.run(Thread.java:1516)

01-Adir-04:426.569(Info) RoverMission : [Danube] Argonaut - Membership Checked.
01-Adir-04:426.569(Info) RoverMission : [Danube] Argonaut - All Boarded.
01-Adir-04:426.569(Info) RoverMission : [Danube] Argonaut - Fulfilled all requirements, departing Danube.
01-Adir-04:426.569(Info) RoverMission : [25.2143 N 95.2903 E] Argonaut - Just embarked from Danube.
01-Adir-04:426.569(Info) RoverMission : [25.2143 N 95.2903 E - Argonaut] Boglarka Kovacs - Just transferred from the settlement to Argonaut.
01-Adir-04:426.569(Info) RoverMission : [25.2143 N 95.2903 E - Argonaut] Sandor Kiss - Just transferred from the settlement to Argonaut.
01-Adir-04:426.569(Info) RoverMission : [25.2143 N 95.2903 E - Argonaut] Jozsef Papp - Just transferred from the settlement to Argonaut.
01-Adir-04:426.569(Warning) Unit : MedicBot 012 expected pulse #8094 but received #8095.

```

**Comment by bevans2000** [comment]
Have you reloaded a saved simulation? Looks like your classpath is corrupted

**Comment by mokun** [comment]
> reloaded a saved simulation? 

I may have changed something in resource process related class. But it shouldn't affect calling `ManufactureWorkTask`...

> Looks like your classpath is corrupted

So how do you tell ?

**Comment by mokun** [comment]
I was running a default scenario with 16 settlements. I have my Monitor Tool with Settlement tab open. It's at 1944x.

I left the simulation alone running at the background.

I was changing a param on a Eclipse launcher file. I never touch any codes in any class though.

Then all of a sudden, I just had another ClassCastException on sol 4. 

So this must have to do with the `MemoryMetricManager` itself.

Can you take a look at what went wrong ? 

It said about calling triggerMalfunction in MalfunctionManager at line 486.

Line 486 : `Simulation.instance().getMetricManager().addValue(entity.getAssociatedSettlement(),
							MALFUNCTION_CAT, OCCURRED_MEASURE, 1);`

```java
/**
	 * Triggers a particular malfunction.
	 *
	 * @param {@link MalfunctionMeta}
	 * @param registerEvent
	 * @param actor
	 */
	public Malfunction triggerMalfunction(MalfunctionMeta m, boolean registerEvent, Unit actor) {
		if (noFailures) {
			return null;
		}
		
		Malfunction malfunction = new Malfunction(this, factory.getNewIncidentNum(), m, supportInsideRepair);

		malfunctions.add(malfunction);
		
		// Update metric
		Simulation.instance().getMetricManager().addValue(entity.getAssociatedSettlement(),
							MALFUNCTION_CAT, OCCURRED_MEASURE, 1);

		numberMalfunctions++;

		getUnit().fireUnitUpdate(MalfunctionManager.MALFUNCTION_EVENT, malfunction);

		if (registerEvent) {
			registerAMalfunction(malfunction, actor);
		}

		if (malfunction.getRepairParts().isEmpty()) { 
			logger.info(actor, 0, malfunction.getName() + " triggered repair work but with no repair parts needed.");	
		}
		
		else {
			calculateNewReliability(malfunction);
		}

		if (!malfunction.getTraumatized().equalsIgnoreCase("None"))
			issueMedicalComplaints(malfunction);

		return malfunction;
	}
```

```
1-Adir-04:419.035(Info) ExploreSite : [15.1325 N 114.1087 E - Sand Spirit] Andy Muller - Expected to collect 23.5 kg rocks to a specimen box.
01-Adir-04:419.035(Info) Exploration : [15.1325 N 114.1087 E - Sand Spirit] Andy Muller - Ready to explore site and collect rocks.
01-Adir-04:419.998(Info) CollectResources : [15.1111 N 114.0320 E - Wind Runner] Ji-min Sun - Collect Ice 3-9 - Collection rate: 2.9
01-Adir-04:422.424(Warning) Building : [Terminus] Solar Thermal Collector 4 - Meteorite Impact event observed. Building wall penetrated.
01-Adir-04:422.424(Info) Malfunction : [Terminus] Solar Thermal Collector 4 - Meteorite Impact Damage - Incident #4. EVA repair triggered. Estimated work time: 43.1 millisols.
01-Adir-04:422.424(Severe) BuildingManager : [Terminus] Solar Thermal Collector 4 - Problem applying pulse to Building
java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1995)
	at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:2111)
	at java.base/java.util.HashMap.treeifyBin(HashMap.java:779)
	at java.base/java.util.HashMap.putVal(HashMap.java:651)
	at java.base/java.util.HashMap.put(HashMap.java:619)
	at com.mars_sim.core.metrics.memory.MemoryMetricManager.getMetric(MemoryMetricManager.java:170)
	at com.mars_sim.core.metrics.MetricManager.getMetric(MetricManager.java:70)
	at com.mars_sim.core.metrics.MetricManager.addValue(MetricManager.java:119)
	at com.mars_sim.core.malfunction.MalfunctionManager.triggerMalfunction(MalfunctionManager.java:486)
	at com.mars_sim.core.building.Building.checkImpactSeverity(Building.java:1272)
	at com.mars_sim.core.building.Building.timePassing(Building.java:1177)
	at com.mars_sim.core.building.BuildingManager.timePassing(BuildingManager.java:944)
	at com.mars_sim.core.structure.Settlement.timePassing(Settlement.java:979)
	at com.mars_sim.core.unit.TemporalThreadExecutor$TemporalRunnable.run(TemporalThreadExecutor.java:81)
	at java.base/java.lang.Thread.run(Thread.java:1516)

01-Adir-04:422.913(Info) GoodsManager : Catarina - Ready to inject Demand for Food: 13.56 -> 16.14  delta: 2.59  fraction: 0.1909  stored: 202.7  reserve: 100.0  optimal: 400.0  lacking: 0.0  surplus: 2.7.
01-Adir-04:422.913(Info) BudgetResources : [Catarina] Arthur Batista - Injected demand for Food: 13.56 -> 15.56
01-Adir-04:423.891(Info) ExploreSite : [9.8069 N 24.7148 E] Pyotr Aksyonov - Collected 1.34 kg Rock samples into a specimen box.
01-Adir-04:423.891(Info) ExploreSite : [9.8069 N 24.7148 E] Pyotr Aksyonov - Exploring site at 9.8069 N 24.7148 E. # of estimation made: 243.
01-Adir-04:423.891(Info) ReviewMissionPlan : [Foundation] Luca Weber - Grading Mining 4-19 - Rating: 5.0, Rels: 9, Quals: 4.0, Obj: 0.0, Emer: 0, Site: 230.5, Dist: -2, Lead: 7, Review: 14, Luck: -3; Subtotal: 264.5
01-Adir-04:423.891(Info) ReviewMissionPlan : [Foundation] Luca Weber - Done reviewing Adriana Moreira's Mining 4-19 mission plan.
01-Adir-04:424.381(Info) ReviewMissionPlan : [Foundation] Jakob Moser - Reviewing Collect Ice 4-22 - Rels: 9, Site: 0.3, Review: 2; Subtotal: 11.3
01-Adir-04:424.381(Info) ReviewMissionPlan : [Foundation] Jakob Moser - Done reviewing Lily Leblanc's Collect Ice 4-22 mission plan.
01-Adir-04:425.362(Info) EVAOperation : [Foundation] Lucas Karlsen - Transferring EVA Suit 364 from person back to rover.
```

**Comment by mokun** [comment]
I still have this `ClassCastException` popping up from time to time.

```
01-Adir-06:268.229(Info) ReviewMissionPlan : [Foundation] Alexander Bauer - Reviewing Mining 6-33 - Rels: 4, Site: 784.5, Review: 2; Subtotal: 790.5
01-Adir-06:268.229(Info) ReviewMissionPlan : [Foundation] Alexander Bauer - Done reviewing Frederike Herrmann's Mining 6-33 mission plan.
01-Adir-06:268.229(Info) BudgetResources : [Foundation] Diederik Hoekstra - Submitted new water rationing: 26.0 -> 25.0
01-Adir-06:268.229(Info) ScientificStudy : Ended the invitation phase with 5 collaborative researchers. Started the research work phase.
01-Adir-06:270.737(Info) CollectResourcesMission : [25.1899 N 95.3599 E - Ikarus] Gabor Balogh - Asleep in Ikarus and not to be disturbed.
01-Adir-06:270.737(Severe) BuildingManager : [New Shanghai] Machinery Hab 1 - Problem applying pulse to Building
java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1995)
	at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:2111)
	at java.base/java.util.HashMap.treeifyBin(HashMap.java:779)
	at java.base/java.util.HashMap.putVal(HashMap.java:651)
	at java.base/java.util.HashMap.put(HashMap.java:619)
	at com.mars_sim.core.metrics.memory.MemoryMetricManager.getMetric(MemoryMetricManager.java:170)
	at com.mars_sim.core.metrics.MetricManager.getMetric(MetricManager.java:70)
	at com.mars_sim.core.metrics.MetricGroup.lambda$0(MetricGroup.java:39)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1229)
	at com.mars_sim.core.metrics.MetricGroup.recordValue(MetricGroup.java:38)
	at com.mars_sim.core.structure.Settlement.addOutput(Settlement.java:2869)
	at com.mars_sim.core.process.ProcessItem.deposit(ProcessItem.java:147)
	at com.mars_sim.core.process.ProcessInfo.depositOutputs(ProcessInfo.java:213)
	at com.mars_sim.core.manufacture.WorkshopProcess.depositOutputs(WorkshopProcess.java:170)
	at com.mars_sim.core.manufacture.ManufactureProcess.stopProcess(ManufactureProcess.java:54)
	at com.mars_sim.core.manufacture.WorkshopProcess.checkStillActive(WorkshopProcess.java:127)
	at com.mars_sim.core.manufacture.WorkshopProcess.addProcessTime(WorkshopProcess.java:117)
	at com.mars_sim.core.building.function.Manufacture.lambda$5(Manufacture.java:328)
	at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:905)
	at com.mars_sim.core.building.function.Manufacture.timePassing(Manufacture.java:328)
	at com.mars_sim.core.building.Building.timePassing(Building.java:1170)
	at com.mars_sim.core.building.BuildingManager.timePassing(BuildingManager.java:931)
	at com.mars_sim.core.structure.Settlement.timePassing(Settlement.java:979)
	at com.mars_sim.core.unit.TemporalThreadExecutor$TemporalRunnable.run(TemporalThreadExecutor.java:82)
	at java.base/java.lang.Thread.run(Thread.java:1516)

01-Adir-06:270.737(Info) ExploreSite [x9] : [56.3282 N 172.2284 E - Baptiste] Lucas Durand - Not EVA fit to explore the site.
01-Adir-06:270.737(Info) Walk : [Terminus - Firefly] Annie Reed - Entered rover Firefly inside Garage 1.
01-Adir-06:270.737(Info) Walk : [Foundation - Dust Storm] Jon Martín - Entered rover Dust Storm inside Loading Dock Garage 2.
01-Adir-06:270.737(Info) MineralSite : [System] 72.9198 N 164.2541 E Degree of estimation certainty improved on 26: 4.8 %
01-Adir-06:273.244(Info) Walk : [Terminus - Firefly] Roy Cooper - Entered rover Firefly inside Garage 1.
```

In Settlement Line 2869, it calls dailyResourceOutput.recordValue(),

```java
	/**
	 * Records the daily output.
	 *
	 * @param resource the resource id of the good
	 * @param amount the amount or quantity produced
	 * @param millisols the labor time
	 */
	public void addOutput(int resource, double amount, double millisols) {
		// This is not optimal and needs a better way to record Measures as not only String
		dailyResourceOutput.recordValue(ResourceType.getName(resource), amount, this);
	}
```

**Comment by bevans2000** [comment]
Is this on a saved and loaded simulation?

**Comment by mokun** [comment]
> Is this on a saved and loaded simulation?

No. I started a brand new simulation when I encountered this.

**Comment by bevans2000** [comment]
Closing as the MetricManager is working. There is a lack of Metrics but this is covered but other issues.
